### PR TITLE
feat(rdb): upgrade gorm v2

### PR DESCRIPTION
**Pull request by MaineK00n**
In gorm, v2 has been released, so we will move to v2.

Refs: https://github.com/go-gorm/gorm

### CODE REVIEW

**Comment by MaineK00n** [comment]
GORM v2 was released in August 2020, so this PR supports it.

## Environments
```terminal
$ gost -v // GORM v1
gost 539d142

$ sqlite3 -version
3.31.1 2020-01-27 19:55:54 3bfa9cc97da10598521b342961df8f5f68c7388fa117345eeb516eaa837balt1

$ mysql --version
mysql  Ver 8.0.23-0ubuntu0.20.04.1 for Linux on x86_64 ((Ubuntu))

$ psql --version
psql (PostgreSQL) 12.6 (Ubuntu 12.6-0ubuntu0.20.04.1)
```

## ChangeLog
### general
Changed some of the redhat and microsoft models for Upsert queries generated by GORMv2 in PostgreSQL (added `ID`).
This made them incompatible, so we introduced a schema version.

### How the schema version works
```terminal
$ ./gost_v2 fetch debian --dbpath=gost_v1.sqlite3
INFO[05-14|09:36:04] Initialize Database 
EROR[05-14|09:36:04] Failed to NewDB. Since SchemaVersion is incompatible, delete Database and fetch again 
Failed to NewDB. Since SchemaVersion is incompatible, delete Database and fetch again.

$ ./gost_v1 fetch debian --dbpath=gost_v2.sqlite3
INFO[05-14|09:37:06] Initialize Database 
EROR[05-14|09:37:06] Failed to migrate db.                    err="Failed to migrate. err: duplicate column name: id"
Failed to migrate. err: duplicate column name: id
```

### SQLite3
- `id`: v1:integer primary key autoincrement => v2:integer primary key (GORM sqlite3 driver bug-like)．
  - The auto increment has been removed, but since SQLite3's ID assignment method is maximum ID+1, this should not have much effect.
- GORMv2's filed tag doesn't support bigint, but sqlite3's integer type stores it in the range of 8 bytes depending on the value, so it shouldn't be a problem.
  - INTEGER. The value is a signed integer, stored in 1, 2, 3, 4, 6, or 8 bytes depending on the magnitude of the value. refs: https://www.sqlite.org/datatype3.html

### MySQL
- v1:datetime => v2: datetime(3)
- In some parts of the Microsoft Model, v1:varchar(255) => v2:text

### PostgreSQL
- Implemented to be aligned with MySQL types
- The Foreign Key constraints that were present in v1 will not be generated in v2.

## sample DB size
```terminal
$ ll gost_v*.sqlite3 
-rw-r--r-- 1 mainek00n mainek00n  82M May 13 04:22 gost_v1.sqlite3
-rw-r--r-- 1 mainek00n mainek00n 133M May 13 02:46 gost_v2.sqlite3
```

## The actual schema that was changed
[mysql_v1.dump.txt](https://github.com/knqyf263/gost/files/6475791/mysql_v1.dump.txt)
[mysql_v2.dump.txt](https://github.com/knqyf263/gost/files/6475792/mysql_v2.dump.txt)
[postgresql_v1.dump.txt](https://github.com/knqyf263/gost/files/6475793/postgresql_v1.dump.txt)
[postgresql_v2.dump.txt](https://github.com/knqyf263/gost/files/6475794/postgresql_v2.dump.txt)
[sqlite_v1.dump.txt](https://github.com/knqyf263/gost/files/6475795/sqlite_v1.dump.txt)
[sqlite_v2.dump.txt](https://github.com/knqyf263/gost/files/6475796/sqlite_v2.dump.txt)

**Comment by MaineK00n** [comment]
Add the following corresponding tests to this PR
- [ ] e.GET("/redhat/:release/pkgs/:name/unfixed-cves", getUnfixedCvesRedhat(driver))
- [ ] e.GET("/debian/:release/pkgs/:name/unfixed-cves", getUnfixedCvesDebian(driver))
- [ ] e.GET("/debian/:release/pkgs/:name/fixed-cves", getFixedCvesDebian(driver))

File: cmd/debian.go [hunk 1]
```diff
@@ -1,11 +1,16 @@
...
+	"errors"
+
 	"github.com/inconshreveable/log15"
 	"github.com/knqyf263/gost/db"
 	"github.com/knqyf263/gost/fetcher"
+	"github.com/knqyf263/gost/models"
 	"github.com/spf13/cobra"
 	"github.com/spf13/viper"
+	"golang.org/x/xerrors"
+	"gorm.io/gorm"
```
[the remark this answers is not in the document]
Comment (MaineK00n) [reply cmd/debian.go]: fixed by d9c8b13

File: cmd/microsoft.go [hunk 2]
```diff
@@ -6,8 +6,11 @@ import (
 	"github.com/inconshreveable/log15"
 	"github.com/knqyf263/gost/db"
 	"github.com/knqyf263/gost/fetcher"
+	"github.com/knqyf263/gost/models"
 	"github.com/spf13/cobra"
 	"github.com/spf13/viper"
+	"golang.org/x/xerrors"
+	"gorm.io/gorm"
```
[the remark this answers is not in the document]
Comment (MaineK00n) [reply cmd/microsoft.go]: fixed by d9c8b13

File: cmd/redhatapi.go [hunk 3]
```diff
@@ -1,13 +1,17 @@
...
 	"fmt"
 
 	"github.com/inconshreveable/log15"
 	"github.com/knqyf263/gost/db"
 	"github.com/knqyf263/gost/fetcher"
+	"github.com/knqyf263/gost/models"
 	"github.com/spf13/cobra"
 	"github.com/spf13/viper"
+	"golang.org/x/xerrors"
+	"gorm.io/gorm"
```
[the remark this answers is not in the document]
Comment (MaineK00n) [reply cmd/redhatapi.go]: fixed by d9c8b13

File: db/rdb.go [hunk 4]
```diff
@@ -91,70 +129,49 @@ func (r *RDBDriver) MigrateDB() error {
...
+	case dialectSqlite3:
+		err = r.conn.Table("sqlite_master").Where("type = ?", "table").Count(&count).Error
+	case dialectMysql:
+		err = r.conn.Table("information_schema.tables").Where("table_schema = ?", r.conn.Migrator().CurrentDatabase()).Count(&count).Error
+	case dialectPostgreSQL:
+		err = r.conn.Table("pg_tables").Where("schemaname = ?", "public").Count(&count).Error
+	}
+
+	if count > 0 {
+		return true, err
```
[the remark this answers is not in the document]
Comment (MaineK00n) [reply db/rdb.go]: fixed by d9c8b13

File: cmd/server.go [hunk 5]
```diff
@@ -1,11 +1,16 @@
...
+	"errors"
+
 	"github.com/inconshreveable/log15"
 	"github.com/knqyf263/gost/db"
+	"github.com/knqyf263/gost/models"
 	"github.com/knqyf263/gost/server"
 	"github.com/spf13/cobra"
 	"github.com/spf13/viper"
+	"golang.org/x/xerrors"
+	"gorm.io/gorm"
```
[the remark this answers is not in the document]
Comment (MaineK00n) [reply cmd/server.go]: fixed by d9c8b13

File: cmd/redhat.go [hunk 6]
```diff
@@ -1,12 +1,16 @@
...
+	"errors"
+
 	"github.com/inconshreveable/log15"
 	"github.com/knqyf263/gost/db"
 	"github.com/knqyf263/gost/fetcher"
+	"github.com/knqyf263/gost/models"
 	"github.com/spf13/cobra"
 	"github.com/spf13/viper"
 	"golang.org/x/xerrors"
+	"gorm.io/gorm"
```
[the remark this answers is not in the document]
Comment (MaineK00n) [reply cmd/redhat.go]: fixed by d9c8b13

File: db/rdb.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (MaineK00n) [reply db/rdb.go]: fixed by 410bf09

**Review by kotakanbe — APPROVED** [review]

File: GNUmakefile [hunk 7]
```diff
@@ -68,3 +68,72 @@ cov:
...
+	-docker rm redis-old redis-new
+
+fetch-rdb:
+	integration/gost.old fetch debian --dbpath=integration/gost.old.sqlite3
+	# integration/gost.old fetch redhat --dbpath=integration/gost.old.sqlite3
+	# integration/gost.old fetch microsoft --dbpath=integration/gost.old.sqlite3 --apikey=<APIKEY>
+	
+	integration/gost.new fetch debian --dbpath=integration/gost.new.sqlite3
+	# integration/gost.new fetch redhat --dbpath=integration/gost.new.sqlite3
+	# integration/gost.old fetch microsoft --dbpath=integration/gost.new.sqlite3 --apikey=<APIKEY>
```
Comment (kotakanbe) [inline GNUmakefile:95]: gost.old -> gost.new

File: GNUmakefile [hunk 8]
```diff
@@ -68,3 +68,72 @@ cov:
...
+	docker run --name redis-old -d -p 127.0.0.1:6379:6379 redis
+	docker run --name redis-new -d -p 127.0.0.1:6380:6379 redis
+
+	integration/gost.old fetch debian --dbtype redis --dbpath "redis://127.0.0.1:6379/0"
+	# integration/gost.old fetch redhat --dbtype redis --dbpath "redis://127.0.0.1:6379/0"
+	# integration/gost.old fetch microsoft --dbtype redis --dbpath "redis://127.0.0.1:6379/0" --apikey=<APIKEY>
+
+	integration/gost.new fetch debian --dbtype redis --dbpath "redis://127.0.0.1:6380/0"
+	# integration/gost.new fetch redhat --dbtype redis --dbpath "redis://127.0.0.1:6380/0"
+	# integration/gost.old fetch microsoft --dbtype redis --dbpath "redis://127.0.0.1:6380/0" --apikey=<APIKEY>
```
Comment (kotakanbe) [inline GNUmakefile:107]: gost.old -> gost.new

**Comment by kotakanbe** [comment]
LGTM

**Pull request merged**